From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/gametest/framework/GameTestHelper.java b/net/minecraft/gametest/framework/GameTestHelper.java
index 7c1d06e8663a76f875f69fd739217f4345f2ee85..d26e7009170d9a416fbe9a3e200216991d7d4009 100644
--- a/net/minecraft/gametest/framework/GameTestHelper.java
+++ b/net/minecraft/gametest/framework/GameTestHelper.java
@@ -370,7 +370,7 @@ public class GameTestHelper {
         };
         Connection connection = new Connection(PacketFlow.SERVERBOUND);
         new EmbeddedChannel(connection);
-        this.getLevel().getServer().getPlayerList().placeNewPlayer(connection, player, cookie);
+        if (true) throw new UnsupportedOperationException(); // Folia - region threading
         return player;
     }
 
