From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/gametest/framework/GameTestServer.java b/net/minecraft/gametest/framework/GameTestServer.java
index 9ca378ed5f29b2928cdb76f13d0eb4b848f6b817..0c19f510f53c13c0851e46dc846b049431091c5c 100644
--- a/net/minecraft/gametest/framework/GameTestServer.java
+++ b/net/minecraft/gametest/framework/GameTestServer.java
@@ -249,8 +249,12 @@ public class GameTestServer extends MinecraftServer {
     }
 
     @Override
-    protected void tickServer(final BooleanSupplier haveTime) {
-        super.tickServer(haveTime);
+    // Folia start - region threading
+    public void tickServer(final long startTime, final long scheduledEnd, final long targetBuffer,
+                           final io.papermc.paper.threadedregions.TickRegions.TickRegionData region) {
+        if (true) throw new UnsupportedOperationException();
+        super.tickServer(startTime, scheduledEnd, targetBuffer, region);
+        // Folia end - region threading
         ServerLevel level = this.overworld();
         if (!this.haveTestsStarted()) {
             this.startTests(level);
