From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/entity/projectile/hurtingprojectile/SmallFireball.java b/net/minecraft/world/entity/projectile/hurtingprojectile/SmallFireball.java
index 55000ce5010321323bf79007ef8de637a492657f..1a72ee1559f75660796a9cbc1a9d9cdd6883dc90 100644
--- a/net/minecraft/world/entity/projectile/hurtingprojectile/SmallFireball.java
+++ b/net/minecraft/world/entity/projectile/hurtingprojectile/SmallFireball.java
@@ -24,7 +24,7 @@ public class SmallFireball extends Fireball {
     public SmallFireball(final Level level, final LivingEntity mob, final Vec3 direction) {
         super(EntityType.SMALL_FIREBALL, mob, direction, level);
         // CraftBukkit start
-        if (this.getOwner() != null && this.getOwner() instanceof Mob) {
+        if (mob != null && mob instanceof Mob) { // Folia - region threading
             this.isIncendiary = (level instanceof ServerLevel serverLevel) && serverLevel.getGameRules().get(GameRules.MOB_GRIEFING);
         }
         // CraftBukkit end
