From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/server/commands/FetchProfileCommand.java b/net/minecraft/server/commands/FetchProfileCommand.java
index 8fdd8b2ef3e07341719d44cc829fc1c4dbd28306..7b9e91cf52b764122c98fea3e2a1dbb3c5953bc9 100644
--- a/net/minecraft/server/commands/FetchProfileCommand.java
+++ b/net/minecraft/server/commands/FetchProfileCommand.java
@@ -126,7 +126,7 @@ public class FetchProfileCommand {
                 () -> {
                     Component nameComponent = Component.literal(name);
                     Optional<GameProfile> result = resolver.fetchByName(name);
-                    server.execute(
+                    io.papermc.paper.threadedregions.RegionizedServer.getInstance().addTask( // Folia - region threading
                         () -> result.ifPresentOrElse(
                             profile -> reportResolvedProfile(source, profile, "commands.fetchprofile.name.success", nameComponent),
                             () -> source.sendFailure(Component.translatable("commands.fetchprofile.name.failure", nameComponent))
@@ -145,7 +145,7 @@ public class FetchProfileCommand {
                 () -> {
                     Component idComponent = Component.translationArg(id);
                     Optional<GameProfile> result = resolver.fetchById(id);
-                    server.execute(
+                    io.papermc.paper.threadedregions.RegionizedServer.getInstance().addTask( // Folia - region threading
                         () -> result.ifPresentOrElse(
                             profile -> reportResolvedProfile(source, profile, "commands.fetchprofile.id.success", idComponent),
                             () -> source.sendFailure(Component.translatable("commands.fetchprofile.id.failure", idComponent))
