From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/server/commands/FillCommand.java b/net/minecraft/server/commands/FillCommand.java
index c8c8ee85004d965f77a00edb033e1863fc03833a..96a2ea1eb51d4f538ed8c8d8a7137de321060f95 100644
--- a/net/minecraft/server/commands/FillCommand.java
+++ b/net/minecraft/server/commands/FillCommand.java
@@ -155,6 +155,12 @@ public class FillCommand {
             );
     }
 
+    // Folia start - region threading
+    private static void sendMessage(CommandSourceStack src, CommandSyntaxException ex) {
+        src.sendFailure((Component)ex.getRawMessage());
+    }
+    // Folia end - region threading
+
     private static int fillBlocks(
         final CommandSourceStack source,
         final BoundingBox region,
@@ -178,6 +184,18 @@ public class FillCommand {
             throw ERROR_FAILED.create();
         }
 
+        // Folia start - region threading
+        int buffer = 32;
+        // physics may spill into neighbour chunks, so use a buffer
+        level.moonrise$loadChunksAsync(
+                (region.minX() - buffer) >> 4,
+                (region.maxX() + buffer) >> 4,
+                (region.minZ() - buffer) >> 4,
+                (region.maxZ() + buffer) >> 4,
+                net.minecraft.world.level.chunk.status.ChunkStatus.FULL,
+                ca.spottedleaf.concurrentutil.util.Priority.NORMAL,
+                (chunks) -> {
+                    try { // Folia end - region threading
         int count = 0;
 
         for (BlockPos pos : BlockPos.betweenClosed(region.minX(), region.minY(), region.minZ(), region.maxX(), region.maxY(), region.maxZ())) {
@@ -219,7 +237,11 @@ public class FillCommand {
 
         int finalCount = count;
         source.sendSuccess(() -> Component.translatable("commands.fill.success", finalCount), true);
-        return count;
+        return; // Folia - region threading
+        // Folia start - region threading
+        } catch (CommandSyntaxException ex) {
+            sendMessage(source, ex);
+        }}); return 0; // Folia end - region threading
     }
 
     @FunctionalInterface
