From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/server/commands/KillCommand.java b/net/minecraft/server/commands/KillCommand.java
index ceef214b551114ba030f0f4b4aaf28aaa8013a3f..a4cfbbbe2a2c3219c0dc996f7b8acd5a2be20d39 100644
--- a/net/minecraft/server/commands/KillCommand.java
+++ b/net/minecraft/server/commands/KillCommand.java
@@ -21,7 +21,9 @@ public class KillCommand {
 
     private static int kill(final CommandSourceStack source, final Collection<? extends Entity> victims) {
         for (Entity entity : victims) {
-            entity.kill(source.getLevel());
+            entity.getBukkitEntity().taskScheduler.scheduleOrExecute((Entity newEntity) -> { // Folia - region threading
+                newEntity.kill(source.getLevel()); // Folia - region threading
+            }); // Folia - region threading
         }
 
         if (victims.size() == 1) {
