From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/server/commands/RecipeCommand.java b/net/minecraft/server/commands/RecipeCommand.java
index fc752a4a8e8f30be0956dd5fc7ef8f08131fd683..37014e8da22b7cd72ca05988d7cbcc537b63d6b7 100644
--- a/net/minecraft/server/commands/RecipeCommand.java
+++ b/net/minecraft/server/commands/RecipeCommand.java
@@ -81,7 +81,12 @@ public class RecipeCommand {
         int success = 0;
 
         for (ServerPlayer player : players) {
-            success += player.awardRecipes(recipes);
+            // Folia start - region threading
+            ++success;
+            player.getBukkitEntity().taskScheduler.schedule((ServerPlayer newPlayer) -> {
+                newPlayer.awardRecipes(recipes);
+            }, null, 1L);
+            // Folia end - region threading
         }
 
         if (success == 0) {
@@ -103,7 +108,12 @@ public class RecipeCommand {
         int success = 0;
 
         for (ServerPlayer player : players) {
-            success += player.resetRecipes(recipes);
+            // Folia start - region threading
+            ++success;
+            player.getBukkitEntity().taskScheduler.schedule((ServerPlayer newPlayer) -> {
+                newPlayer.resetRecipes(recipes);
+            }, null, 1L);
+            // Folia end - region threading
         }
 
         if (success == 0) {
