From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/server/commands/TeleportCommand.java b/net/minecraft/server/commands/TeleportCommand.java
index 290547abf7037c5c88736c66765b2ee8f575c53b..369be2ad59cf05c76a44b4b00bec18d1c564688a 100644
--- a/net/minecraft/server/commands/TeleportCommand.java
+++ b/net/minecraft/server/commands/TeleportCommand.java
@@ -150,18 +150,7 @@ public class TeleportCommand {
 
     private static int teleportToEntity(final CommandSourceStack source, final Collection<? extends Entity> entities, final Entity destination) throws CommandSyntaxException {
         for (Entity entity : entities) {
-            performTeleport(
-                source,
-                entity,
-                (ServerLevel)destination.level(),
-                destination.getX(),
-                destination.getY(),
-                destination.getZ(),
-                EnumSet.noneOf(Relative.class),
-                destination.getYRot(),
-                destination.getXRot(),
-                null
-            );
+            io.papermc.paper.threadedregions.TeleportUtils.teleport(entity, false, destination, Float.valueOf(destination.getYRot()), Float.valueOf(destination.getXRot()), Entity.TELEPORT_FLAG_LOAD_CHUNK, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.COMMAND, null); // Folia - region threading
         }
 
         if (entities.size() == 1) {
@@ -256,6 +245,23 @@ public class TeleportCommand {
         float relativeOrAbsoluteXRot = relatives.contains(Relative.X_ROT) ? xRot - victim.getXRot() : xRot;
         float newYRot = Mth.wrapDegrees(relativeOrAbsoluteYRot);
         float newXRot = Mth.wrapDegrees(relativeOrAbsoluteXRot);
+        // Folia start - region threading
+        if (true) {
+            Vec3 posFinal = new Vec3(x, y, z);
+            Float yawFinal = Float.valueOf(newYRot);
+            Float pitchFinal = Float.valueOf(newXRot);
+            victim.getBukkitEntity().taskScheduler.schedule((Entity nmsEntity) -> {
+                nmsEntity.stopRiding();
+                nmsEntity.teleportAsync(
+                        level, posFinal, yawFinal, pitchFinal, Vec3.ZERO,
+                        org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.COMMAND,
+                        Entity.TELEPORT_FLAG_LOAD_CHUNK | Entity.TELEPORT_FLAG_TELEPORT_PASSENGERS,
+                        null
+                );
+                }, null, 1L);
+            return;
+        }
+        // Folia end - region threading
         if (victim.teleportTo(level, relativeOrAbsoluteX, relativeOrAbsoluteY, relativeOrAbsoluteZ, relatives, newYRot, newXRot, true, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.COMMAND)) { // Paper - teleport cause
             if (lookAt != null) {
                 lookAt.perform(source, victim);
