From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/entity/raid/Raider.java b/net/minecraft/world/entity/raid/Raider.java
index cf689dae6978577970e39bf3004be551b4908cd1..26d4213691a30741b7f87ce1d039e7e0340a4185 100644
--- a/net/minecraft/world/entity/raid/Raider.java
+++ b/net/minecraft/world/entity/raid/Raider.java
@@ -89,7 +89,7 @@ public abstract class Raider extends PatrollingMonster {
             Raid currentRaid = this.getCurrentRaid();
             if (this.canJoinRaid()) {
                 if (currentRaid == null) {
-                    if (this.level().getGameTime() % 20L == 0L) {
+                    if (this.level().getRedstoneGameTime() % 20L == 0L) { // Folia - region threading
                         Raid nearbyRaid = level.getRaidAt(this.blockPosition());
                         if (nearbyRaid != null && Raids.canJoinRaid(this)) {
                             nearbyRaid.joinRaid(level, nearbyRaid.getGroupsSpawned(), this, null, true);
