--- a/net/minecraft/world/entity/npc/villager/AbstractVillager.java
+++ b/net/minecraft/world/entity/npc/villager/AbstractVillager.java
@@ -213,9 +_,17 @@
         this.readInventoryFromTag(input);
     }
 
+    // Folia start - region threading
     @Override
-    public @Nullable Entity teleport(final TeleportTransition transition) {
+    public void preChangeDimension() {
+        super.preChangeDimension();
         this.stopTrading();
+    }
+    // Folia end - region threading
+
+    @Override
+    public @Nullable Entity teleport(final TeleportTransition transition) {
+        this.preChangeDimension(); // Folia - region threading - move into preChangeDimension
         return super.teleport(transition);
     }
 
