From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/level/saveddata/SavedData.java b/net/minecraft/world/level/saveddata/SavedData.java
index 78360fb439aa3b3b8702cc293378849f4190d364..ed576f17f5901b79350aca84fd56fa996f55e258 100644
--- a/net/minecraft/world/level/saveddata/SavedData.java
+++ b/net/minecraft/world/level/saveddata/SavedData.java
@@ -1,7 +1,7 @@
 package net.minecraft.world.level.saveddata;
 
 public abstract class SavedData {
-    private boolean dirty;
+    private volatile boolean dirty; // Folia - make map data thread-safe
 
     public void setDirty() {
         this.setDirty(true);
