From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/level/BaseCommandBlock.java b/net/minecraft/world/level/BaseCommandBlock.java
index fbe1cf1f02a27cbe0666fb83eb74c82df92785c9..31f1583f29eebe5e13b5a7ac5116fdafc3592e5c 100644
--- a/net/minecraft/world/level/BaseCommandBlock.java
+++ b/net/minecraft/world/level/BaseCommandBlock.java
@@ -91,6 +91,7 @@ public abstract class BaseCommandBlock {
     }
 
     public boolean performCommand(final ServerLevel level) {
+        if (true) return false; // Folia - region threading
         if (level.getGameTime() == this.lastExecution) {
             return false;
         }
@@ -173,9 +174,11 @@ public abstract class BaseCommandBlock {
 
     public abstract boolean isValid();
 
+    public void threadCheck() {} // Folia
+
     public class CloseableCommandBlockSource implements CommandSource, AutoCloseable { // Paper - public
         private final ServerLevel level;
-        private static final DateTimeFormatter TIME_FORMAT = DateTimeFormatter.ofPattern("HH:mm:ss", Locale.ROOT);
+        private static final ThreadLocal<DateTimeFormatter> TIME_FORMAT = ThreadLocal.withInitial(() -> DateTimeFormatter.ofPattern("HH:mm:ss", Locale.ROOT)); // Folia - region threading - SDF is not thread-safe
         private boolean closed;
 
         // Paper start - add back source when output disabled
@@ -209,7 +212,8 @@ public abstract class BaseCommandBlock {
         public void sendSystemMessage(final Component message) {
             if (this.trackOutput && !this.closed) { // Paper - add back source when output disabled
                 org.spigotmc.AsyncCatcher.catchOp("sendSystemMessage to a command block"); // Paper - Don't broadcast messages to command blocks
-                BaseCommandBlock.this.lastOutput = Component.literal("[" + TIME_FORMAT.format(ZonedDateTime.now()) + "] ").append(message);
+                BaseCommandBlock.this.threadCheck(); // Folia
+                BaseCommandBlock.this.lastOutput = Component.literal("[" + TIME_FORMAT.get().format(ZonedDateTime.now()) + "] ").append(message); // Folia - region threading - SDF is not thread-safe
                 BaseCommandBlock.this.onUpdated(this.level);
             }
         }
