From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/level/StructureManager.java b/net/minecraft/world/level/StructureManager.java
index a8f00da4d5c7cca9b6948d7c2d5114f84fdacc00..3e6f66f44d3c2518879d57dc16a531e5edf1805b 100644
--- a/net/minecraft/world/level/StructureManager.java
+++ b/net/minecraft/world/level/StructureManager.java
@@ -48,13 +48,7 @@ public class StructureManager {
     }
 
     public List<StructureStart> startsForStructure(final ChunkPos pos, final Predicate<Structure> matcher) {
-        // Paper start - Fix swamp hut cat generation deadlock
-        return this.startsForStructure(pos, matcher, null);
-    }
-
-    public List<StructureStart> startsForStructure(final ChunkPos pos, final Predicate<Structure> matcher, final @Nullable ServerLevelAccessor levelAccessor) {
-        Map<Structure, LongSet> allReferences = (levelAccessor == null ? this.level : levelAccessor).getChunk(pos.x(), pos.z(), ChunkStatus.STRUCTURE_REFERENCES).getAllReferences();
-        // Paper end - Fix swamp hut cat generation deadlock
+        Map<Structure, LongSet> allReferences = this.level.getChunk(pos.x(), pos.z(), ChunkStatus.STRUCTURE_REFERENCES).getAllReferences(); // Folia - region threading
         Builder<StructureStart> result = ImmutableList.builder();
 
         for (Entry<Structure, LongSet> entry : allReferences.entrySet()) {
@@ -117,29 +111,15 @@ public class StructureManager {
     }
 
     public StructureStart getStructureWithPieceAt(final BlockPos blockPos, final HolderSet<Structure> structures) {
-        // Paper start - Fix swamp hut cat generation deadlock
-        return this.getStructureWithPieceAt(blockPos, structures, null);
-    }
-    public StructureStart getStructureWithPieceAt(final BlockPos blockPos, final HolderSet<Structure> structures, final @Nullable ServerLevelAccessor levelAccessor) {
-        return this.getStructureWithPieceAt(blockPos, structures::contains, levelAccessor);
-        // Paper end - Fix swamp hut cat generation deadlock
+        return this.getStructureWithPieceAt(blockPos, structures::contains); // Folia - region threading
     }
 
     public StructureStart getStructureWithPieceAt(final BlockPos blockPos, final Predicate<Holder<Structure>> predicate) {
-        // Paper start - Fix swamp hut cat generation deadlock
-        return this.getStructureWithPieceAt(blockPos, predicate, null);
-    }
-
-    public StructureStart getStructureWithPieceAt(final BlockPos blockPos, final TagKey<Structure> structureTag, final @Nullable ServerLevelAccessor levelAccessor) {
-        return this.getStructureWithPieceAt(blockPos, structure -> structure.is(structureTag), levelAccessor);
-    }
-
-    public StructureStart getStructureWithPieceAt(final BlockPos blockPos, final Predicate<Holder<Structure>> predicate, final @Nullable ServerLevelAccessor levelAccessor) {
-        // Paper end - Fix swamp hut cat generation deadlock
+        // Folia - region threading
         Registry<Structure> structures = this.registryAccess().lookupOrThrow(Registries.STRUCTURE);
 
         for (StructureStart structureStart : this.startsForStructure(
-            ChunkPos.containing(blockPos), s -> structures.get(structures.getId(s)).map(predicate::test).orElse(false), levelAccessor // Paper - Fix swamp hut cat generation deadlock
+            ChunkPos.containing(blockPos), s -> structures.get(structures.getId(s)).map(predicate::test).orElse(false) // Paper - Fix swamp hut cat generation deadlock // Folia - region threading
         )) {
             if (this.structureHasPieceAt(blockPos, structureStart)) {
                 return structureStart;
@@ -186,7 +166,7 @@ public class StructureManager {
     }
 
     public void addReference(final StructureStart start) {
-        start.addReference();
+        //start.addReference(); // Folia - region threading - move to caller
         this.structureCheck.incrementReference(start.getChunkPos(), start.getStructure());
     }
 
