--- a/net/minecraft/server/commands/DamageCommand.java
+++ b/net/minecraft/server/commands/DamageCommand.java
@@ -103,12 +_,29 @@
         );
     }
 
+    // Folia start - region threading
+    private static void sendMessage(CommandSourceStack src, CommandSyntaxException ex) {
+        src.sendFailure((Component)ex.getRawMessage());
+    }
+    // Folia end - region threading
+
     private static int damage(final CommandSourceStack stack, final Entity target, final float amount, final DamageSource source) throws CommandSyntaxException {
-        if (target.hurtServer(stack.getLevel(), source, amount)) {
-            stack.sendSuccess(() -> Component.translatable("commands.damage.success", amount, target.getDisplayName()), true);
-            return Command.SINGLE_SUCCESS;
+        // Folia start - region threading
+        target.getBukkitEntity().taskScheduler.scheduleOrExecute((Entity newTarget) -> {
+            try {
+                // Folia end - region threading
+        if (newTarget.hurtServer(stack.getLevel(), source, amount)) { // Folia - region threading
+            stack.sendSuccess(() -> Component.translatable("commands.damage.success", amount, newTarget.getDisplayName()), true); // Folia - region threading
+            return; // Folia - region threading
         } else {
             throw ERROR_INVULNERABLE.create();
         }
+        // Folia start - region threading
+        } catch (CommandSyntaxException ex) {
+            sendMessage(stack, ex);
+        }
+        });
+        return 0;
+        // Folia end - region threading
     }
 }
