--- a/net/minecraft/server/commands/EffectCommands.java
+++ b/net/minecraft/server/commands/EffectCommands.java
@@ -178,7 +_,11 @@
         for (Entity entity : entities) {
             if (entity instanceof LivingEntity livingEntity) {
                 MobEffectInstance instance = new MobEffectInstance(effectHolder, duration, amplifier, false, particles);
-                if (livingEntity.addEffect(instance, source.getEntity(), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.COMMAND)) { // CraftBukkit
+                entity.getBukkitEntity().taskScheduler.scheduleOrExecute((LivingEntity newEntity) -> {
+                    newEntity.addEffect(instance, source.getEntity(), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.COMMAND);
+                });
+                // Folia end - region threading
+                if (true) { // CraftBukkit // CraftBukkit // Folia - region threading
                     count++;
                 }
             }
@@ -208,7 +_,12 @@
         int count = 0;
 
         for (Entity entity : entities) {
-            if (entity instanceof LivingEntity livingEntity && livingEntity.removeAllEffects(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.COMMAND)) { // CraftBukkit
+            if (entity instanceof LivingEntity && true) { // CraftBukkit // Folia - region threading
+                // Folia start - region threading
+                entity.getBukkitEntity().taskScheduler.scheduleOrExecute((LivingEntity newEntity) -> {
+                    newEntity.removeAllEffects(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.COMMAND);
+                });
+                // Folia end - region threading
                 count++;
             }
         }
@@ -233,7 +_,12 @@
         int count = 0;
 
         for (Entity entity : entities) {
-            if (entity instanceof LivingEntity livingEntity && livingEntity.removeEffect(effectHolder, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.COMMAND)) { // CraftBukkit
+            if (entity instanceof LivingEntity livingEntity && true) { // CraftBukkit // Folia - region threading
+                // Folia start - region threading
+                entity.getBukkitEntity().taskScheduler.scheduleOrExecute((LivingEntity newEntity) -> {
+                    newEntity.removeEffect(effectHolder, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.COMMAND);
+                });
+                // Folia end - region threading
                 count++;
             }
         }
