--- a/net/minecraft/server/commands/ExperienceCommand.java
+++ b/net/minecraft/server/commands/ExperienceCommand.java
@@ -123,16 +_,23 @@
     }
 
     private static int queryExperience(final CommandSourceStack source, final ServerPlayer target, final ExperienceCommand.Type type) {
-        int result = type.query.applyAsInt(target);
-        source.sendSuccess(() -> Component.translatable("commands.experience.query." + type.name, target.getDisplayName(), result), false);
-        return result;
+        // Folia start - region threading
+        target.getBukkitEntity().taskScheduler.scheduleOrExecute((ServerPlayer newTarget) -> {
+            int result = type.query.applyAsInt(newTarget);
+            source.sendSuccess(() -> Component.translatable("commands.experience.query." + type.name, newTarget.getDisplayName(), result), false);
+            return;
+        });
+        return 0;
+        // Folia end - region threading
     }
 
     private static int addExperience(
         final CommandSourceStack source, final Collection<? extends ServerPlayer> players, final int amount, final ExperienceCommand.Type type
     ) {
         for (ServerPlayer player : players) {
-            type.add.accept(player, amount);
+            player.getBukkitEntity().taskScheduler.schedule((ServerPlayer newPlayer) -> { // Folia - region threading
+            type.add.accept(newPlayer, amount); // Folia - region threading
+            }, null, 1L); // Folia - region threading
         }
 
         if (players.size() == 1) {
@@ -153,9 +_,14 @@
         int success = 0;
 
         for (ServerPlayer player : players) {
-            if (type.set.test(player, amount)) {
-                success++;
+            // Folia start - region threading
+            ++success;
+            player.getBukkitEntity().taskScheduler.schedule((ServerPlayer newPlayer) -> {
+            if (type.set.test(newPlayer, amount)) {
+                //success++;
+                // Folia end - region threading
             }
+            }, null, 1L); // Folia - region threading
         }
 
         if (success == 0) {
