--- a/net/minecraft/server/commands/GameModeCommand.java
+++ b/net/minecraft/server/commands/GameModeCommand.java
@@ -56,9 +_,11 @@
                 server.setDefaultGameType(type);
             }
 
-            if (setGameMode(context.getSource(), player, type)) {
-                count++;
+            ++count; player.getBukkitEntity().taskScheduler.scheduleOrExecute((ServerPlayer newPlayer) -> { // Folia - region threading
+            if (setGameMode(context.getSource(), newPlayer, type)) { // Folia - region threading
+                //count++; // Folia - region threading
             }
+            }); // Folia - region threading
         }
 
         return count;
