--- a/net/minecraft/server/commands/GiveCommand.java
+++ b/net/minecraft/server/commands/GiveCommand.java
@@ -61,32 +_,34 @@
                 int size = Math.min(maxStackSize, remaining);
                 remaining -= size;
                 ItemStack copyToDrop = prototypeItemStack.copyWithCount(size);
-                boolean added = player.getInventory().add(copyToDrop);
+                player.getBukkitEntity().taskScheduler.scheduleOrExecute((ServerPlayer newPlayer) -> { // Folia - region threading
+                boolean added = newPlayer.getInventory().add(copyToDrop); // Folia - region threading
                 if (added && copyToDrop.isEmpty()) {
-                    ItemEntity drop = player.drop(prototypeItemStack.copy(), false, false, false, null); // Paper - do not fire PlayerDropItemEvent for /give command
+                    ItemEntity drop = newPlayer.drop(prototypeItemStack.copy(), false, false, false, null); // Paper - do not fire PlayerDropItemEvent for /give command // Folia - region threading
                     if (drop != null) {
                         drop.makeFakeItem();
                     }
 
-                    player.level()
+                    newPlayer.level() // Folia - region threading
                         .playSound(
                             null,
-                            player.getX(),
-                            player.getY(),
-                            player.getZ(),
+                            newPlayer.getX(), // Folia - region threading
+                            newPlayer.getY(), // Folia - region threading
+                            newPlayer.getZ(), // Folia - region threading
                             SoundEvents.ITEM_PICKUP,
                             SoundSource.PLAYERS,
                             0.2F,
-                            ((player.getRandom().nextFloat() - player.getRandom().nextFloat()) * 0.7F + 1.0F) * 2.0F
+                            ((newPlayer.getRandom().nextFloat() - newPlayer.getRandom().nextFloat()) * 0.7F + 1.0F) * 2.0F // Folia - region threading
                         );
-                    player.containerMenu.broadcastChanges();
+                    newPlayer.containerMenu.broadcastChanges(); // Folia - region threading
                 } else {
-                    ItemEntity drop = player.drop(copyToDrop, false, false, false, null); // Paper - do not fire PlayerDropItemEvent for /give command
+                    ItemEntity drop = newPlayer.drop(copyToDrop, false, false, false, null); // Paper - do not fire PlayerDropItemEvent for /give command // Folia - region threading
                     if (drop != null) {
                         drop.setNoPickUpDelay();
-                        drop.setTarget(player.getUUID());
+                        drop.setTarget(newPlayer.getUUID()); // Folia - region threading
                     }
                 }
+                }); // Folia - region threading
             }
         }
 
