--- a/net/minecraft/server/commands/RecipeCommand.java
+++ b/net/minecraft/server/commands/RecipeCommand.java
@@ -81,7 +_,12 @@
         int success = 0;
 
         for (ServerPlayer player : players) {
-            success += player.awardRecipes(recipes);
+            // Folia start - region threading
+            ++success;
+            player.getBukkitEntity().taskScheduler.schedule((ServerPlayer newPlayer) -> {
+                newPlayer.awardRecipes(recipes);
+            }, null, 1L);
+            // Folia end - region threading
         }
 
         if (success == 0) {
@@ -103,7 +_,12 @@
         int success = 0;
 
         for (ServerPlayer player : players) {
-            success += player.resetRecipes(recipes);
+            // Folia start - region threading
+            ++success;
+            player.getBukkitEntity().taskScheduler.schedule((ServerPlayer newPlayer) -> {
+                newPlayer.resetRecipes(recipes);
+            }, null, 1L);
+            // Folia end - region threading
         }
 
         if (success == 0) {
