--- a/net/minecraft/server/commands/SummonCommand.java
+++ b/net/minecraft/server/commands/SummonCommand.java
@@ -98,11 +_,17 @@
             throw ERROR_FAILED.create();
         }
 
+        // Folia start - region threading
+        io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueOrExecuteTickTask(
+                level, entity.chunkPosition().x(), entity.chunkPosition().z(), () -> {
+                    // Folia end - region threading
         if (finalize && entity instanceof Mob mob) {
             mob.finalizeSpawn(source.getLevel(), source.getLevel().getCurrentDifficultyAt(entity.blockPosition()), EntitySpawnReason.COMMAND, null);
         }
+        level.tryAddFreshEntityWithPassengers(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.COMMAND); // Folia - region threading
+        }); // Folia - region threading
 
-        if (!level.tryAddFreshEntityWithPassengers(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.COMMAND)) { // CraftBukkit - pass a spawn reason of "COMMAND"
+        if (false) { // CraftBukkit - pass a spawn reason of "COMMAND" // Folia - region threading
             throw ERROR_DUPLICATE_UUID.create();
         } else {
             return entity;
