From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java b/net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java
index 65bbb51f13e825270573c5402dbc537731aa3dde..db3e0bc9defb2f1ea424346adfab2da6c9f5298b 100644
--- a/net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java
+++ b/net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java
@@ -45,12 +45,14 @@ public class GoToPotentialJobSite extends Behavior<Villager> {
             BlockPos pos = globalPos.pos();
             ServerLevel serverLevel = level.getServer().getLevel(globalPos.dimension());
             if (serverLevel != null) {
+                io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueOrExecuteTickTask(serverLevel, pos.getX() >> 4, pos.getZ() >> 4, () -> { // Folia - region threading
                 PoiManager manager = serverLevel.getPoiManager();
                 if (manager.exists(pos, p -> true)) {
                     manager.release(pos);
                 }
 
                 level.debugSynchronizers().updatePoi(pos);
+                }); // Folia - region threading
             }
         });
         body.getBrain().eraseMemory(MemoryModuleType.POTENTIAL_JOB_SITE);
