From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/server/level/ServerEntityGetter.java b/net/minecraft/server/level/ServerEntityGetter.java
index 3a742c3ae5c32701a0934f86c0c33cd50f316652..4554d7a112f5e8968fa02af8f49a4b74158031bf 100644
--- a/net/minecraft/server/level/ServerEntityGetter.java
+++ b/net/minecraft/server/level/ServerEntityGetter.java
@@ -15,17 +15,17 @@ public interface ServerEntityGetter extends EntityGetter {
     ServerLevel getLevel();
 
     default @Nullable Player getNearestPlayer(final TargetingConditions targetConditions, final LivingEntity source) {
-        return this.getNearestEntity(this.players(), targetConditions, source, source.getX(), source.getY(), source.getZ());
+        return this.getNearestEntity(this.getLocalPlayers(), targetConditions, source, source.getX(), source.getY(), source.getZ()); // Folia - region threading
     }
 
     default @Nullable Player getNearestPlayer(
         final TargetingConditions targetConditions, final LivingEntity source, final double x, final double y, final double z
     ) {
-        return this.getNearestEntity(this.players(), targetConditions, source, x, y, z);
+        return this.getNearestEntity(this.getLocalPlayers(), targetConditions, source, x, y, z); // Folia - region threading
     }
 
     default @Nullable Player getNearestPlayer(final TargetingConditions targetConditions, final double x, final double y, final double z) {
-        return this.getNearestEntity(this.players(), targetConditions, null, x, y, z);
+        return this.getNearestEntity(this.getLocalPlayers(), targetConditions, null, x, y, z); // Folia - region threading
     }
 
     default <T extends LivingEntity> @Nullable T getNearestEntity(
@@ -92,7 +92,7 @@ public interface ServerEntityGetter extends EntityGetter {
     default List<Player> getNearbyPlayers(final TargetingConditions targetConditions, final LivingEntity source, final AABB bb) {
         List<Player> foundPlayers = new ArrayList<>();
 
-        for (Player player : this.players()) {
+        for (Player player : this.getLocalPlayers()) { // Folia - region threading
             if (bb.contains(player.getX(), player.getY(), player.getZ()) && targetConditions.test(this.getLevel(), source, player)) {
                 foundPlayers.add(player);
             }
