From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/server/players/OldUsersConverter.java b/net/minecraft/server/players/OldUsersConverter.java
index b0acf291dca0d6f50b5cfc36cb6ed33a369fe5c2..09ae4ac8235b60612ca9dbc62e51266e2bdf6934 100644
--- a/net/minecraft/server/players/OldUsersConverter.java
+++ b/net/minecraft/server/players/OldUsersConverter.java
@@ -445,7 +445,7 @@ public class OldUsersConverter {
     private static Date parseDate(final String dateString, final Date defaultValue) {
         Date parsedDate;
         try {
-            parsedDate = BanListEntry.DATE_FORMAT.parse(dateString);
+            parsedDate = BanListEntry.DATE_FORMAT.get().parse(dateString); // Folia - region threading - SDF is not thread-safe
         } catch (ParseException ignored) {
             parsedDate = defaultValue;
         }
