From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java b/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
index 471b33333c32e49076fd4cc169d01da4a92f3101..22ad3d2f6cc28027e581fd998d41b78f20b46b20 100644
--- a/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
@@ -54,7 +54,7 @@ public class EndCrystal extends Entity {
     public void tick() {
         this.time++;
         this.applyEffectsFromBlocks();
-        this.handlePortal();
+        //this.handlePortal(); // Folia - region threading
         if (this.level() instanceof ServerLevel) {
             BlockPos pos = this.blockPosition();
             if (((ServerLevel)this.level()).getDragonFight() != null && this.level().getBlockState(pos).isAir()) {
