From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/entity/ai/sensing/TemptingSensor.java b/net/minecraft/world/entity/ai/sensing/TemptingSensor.java
index f0dc4231b79e4f02416917dacce978121ec6d7e2..f5a1d35d41167d794409b4a585a7dc0f16ff8d96 100644
--- a/net/minecraft/world/entity/ai/sensing/TemptingSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/TemptingSensor.java
@@ -38,7 +38,7 @@ public class TemptingSensor extends Sensor<PathfinderMob> {
     protected void doTick(final ServerLevel level, final PathfinderMob body) {
         Brain<?> brain = body.getBrain();
         TargetingConditions targeting = TEMPT_TARGETING.copy().range((float)body.getAttributeValue(Attributes.TEMPT_RANGE));
-        List<Player> players = level.players()
+        List<Player> players = level.getLocalPlayers() // Folia - region threading
             .stream()
             .filter(EntitySelector.NO_SPECTATORS)
             .filter(playerx -> targeting.test(level, body, playerx))
