--- a/net/minecraft/world/clock/ServerClockManager.java
+++ b/net/minecraft/world/clock/ServerClockManager.java
@@ -160,7 +_,7 @@
     }
 
     private long getGameTime() {
-        return this.level != null ? this.level.getGameTime() : this.server.overworld().getGameTime(); // Paper - per-world time
+        return this.level != null ? this.level.levelData.getGameTime() : this.server.overworld().levelData.getGameTime(); // Paper - per-world time // Folia - region threading
     }
 
     // Paper start - per-world time; per-player time
