From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/entity/ai/Brain.java b/net/minecraft/world/entity/ai/Brain.java
index 07a9f94438fd82830c5e53ecb86d4dd6d67631b0..d4af960ac09dc609b89ccc42e3bd59436506ab88 100644
--- a/net/minecraft/world/entity/ai/Brain.java
+++ b/net/minecraft/world/entity/ai/Brain.java
@@ -400,9 +400,17 @@ public class Brain<E extends LivingEntity> {
     }
 
     public void stopAll(final ServerLevel level, final E body) {
+        // Folia start - region threading
+        final List<BehaviorControl<? super E>> behaviors = this.getRunningBehaviors();
+        if (behaviors.isEmpty()) {
+            // avoid calling getGameTime, as this may be called while portalling an entity - which will cause
+            // the world data retrieval to fail
+            return;
+        }
+        // Folia end - region threading
         long timestamp = body.level().getGameTime();
 
-        for (BehaviorControl<? super E> behavior : this.getRunningBehaviors()) {
+        for (BehaviorControl<? super E> behavior : behaviors) { // Folia - region threading
             behavior.doStop(level, body, timestamp);
         }
     }
