From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/entity/decoration/ItemFrame.java b/net/minecraft/world/entity/decoration/ItemFrame.java
index 9cc2af3db862fdfa45a67ef4b9c49581d7d69d36..310ee11ee6b864723f2d2cb15c78e8f01ce1dad9 100644
--- a/net/minecraft/world/entity/decoration/ItemFrame.java
+++ b/net/minecraft/world/entity/decoration/ItemFrame.java
@@ -267,7 +267,9 @@ public class ItemFrame extends HangingEntity {
         if (mapId != null) {
             MapItemSavedData mapItemSavedData = MapItem.getSavedData(mapId, this.level());
             if (mapItemSavedData != null) {
+                synchronized (mapItemSavedData) { // Folia - make map data thread-safe
                 mapItemSavedData.removedFromFrame(this.pos, this.getId());
+                } // Folia - make map data thread-safe
             }
         }
     }
