--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -506,9 +_,20 @@
         return false;
     }
 
+    // Folia start - region threading
+    @Override
+    public void postChangeDimension() {
+        super.postChangeDimension();
+        if (!this.level().isClientSide() && this instanceof ItemEntity item) {
+            item.mergeWithNeighbours();
+        }
+    }
+    // Folia end - region threading
+
     @Override
     public @Nullable Entity teleport(final TeleportTransition transition) {
         Entity entity = super.teleport(transition);
+        if (entity != null) entity.postChangeDimension(); // Folia - region threading - move to post change
         if (!this.level().isClientSide() && entity instanceof ItemEntity item) {
             item.mergeWithNeighbours();
         }
