From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/level/block/entity/BeaconBlockEntity.java b/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
index beca8062d213c010f2135c842cc7b1bdcaf94d6e..c2cf8df9bec98d372637b8c1e4c90444455f437a 100644
--- a/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
@@ -209,7 +209,7 @@ public class BeaconBlockEntity extends BlockEntity implements MenuProvider, Name
         }
 
         int previousLevels = entity.levels;
-        if (level.getGameTime() % 80L == 0L) {
+        if (level.getRedstoneGameTime() % 80L == 0L) { // Folia - region threading
             if (!entity.beamSections.isEmpty()) {
                 entity.levels = updateBase(level, x, y, z);
             }
@@ -340,7 +340,7 @@ public class BeaconBlockEntity extends BlockEntity implements MenuProvider, Name
             list = level.getEntitiesOfClass(Player.class, aabb); // Diff from applyEffect
         } else {
             list = new java.util.ArrayList<>();
-            for (final Player player : level.players()) {
+            for (final Player player : level.getLocalPlayers()) { // Folia - region threading
                 if (!net.minecraft.world.entity.EntitySelector.NO_SPECTATORS.test(player)) continue;
                 if (player.getBoundingBox().intersects(aabb)) {
                     list.add(player);
