--- a/net/minecraft/world/level/chunk/ChunkGenerator.java
+++ b/net/minecraft/world/level/chunk/ChunkGenerator.java
@@ -330,7 +_,7 @@
     }
 
     private static boolean tryAddReference(final StructureManager manager, final StructureStart start) {
-        if (start.canBeReferenced()) {
+        if (start.tryReference()) { // Folia - region threading
             manager.addReference(start);
             return true;
         } else {
