From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/level/block/DaylightDetectorBlock.java b/net/minecraft/world/level/block/DaylightDetectorBlock.java
index 8dd27ab87bb05ccdbf226a6f121c6d5e65a3c0c0..7f7efed5241907c4791d317268d98148e22c07da 100644
--- a/net/minecraft/world/level/block/DaylightDetectorBlock.java
+++ b/net/minecraft/world/level/block/DaylightDetectorBlock.java
@@ -111,7 +111,7 @@ public class DaylightDetectorBlock extends BaseEntityBlock {
     }
 
     private static void tickEntity(final Level level, final BlockPos blockPos, final BlockState blockState, final DaylightDetectorBlockEntity blockEntity) {
-        if (level.getGameTime() % 20L == 0L) {
+        if (level.getRedstoneGameTime() % 20L == 0L) { // Folia - region threading
             updateSignalStrength(blockState, level, blockPos);
         }
     }
