From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/entity/item/FallingBlockEntity.java b/net/minecraft/world/entity/item/FallingBlockEntity.java
index 045ce9cf39ec1c106bf7e9694e6d309f58900850..1cdea4bea79025184b6cc902b33e3e8186b72e89 100644
--- a/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -164,7 +164,7 @@ public class FallingBlockEntity extends Entity {
                 return;
             }
             // Paper end - Configurable falling blocks height nerf
-            this.handlePortal();
+            //this.handlePortal(); // Folia - region threading
             if (this.level() instanceof ServerLevel serverLevel && (this.isAlive() || this.forceTickAfterTeleportToDuplicate)) {
                 BlockPos pos = this.blockPosition();
                 boolean isConcrete = this.blockState.getBlock() instanceof ConcretePowderBlock;
