From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/entity/item/PrimedTnt.java b/net/minecraft/world/entity/item/PrimedTnt.java
index 7970c6f9189059e8fc0391709159ef395c56708c..42aff39d537aa6b138bc4a76304d49f423212acd 100644
--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -101,8 +101,8 @@ public class PrimedTnt extends Entity implements TraceableEntity {
 
     @Override
     public void tick() {
-        if (this.level().spigotConfig.maxTntTicksPerTick > 0 && ++this.level().spigotConfig.currentPrimedTnt > this.level().spigotConfig.maxTntTicksPerTick) { return; } // Spigot
-        this.handlePortal();
+        if (this.level().spigotConfig.maxTntTicksPerTick > 0 && ++this.level().getCurrentWorldData().currentPrimedTnt > this.level().spigotConfig.maxTntTicksPerTick) { return; } // Spigot // Folia - region threading
+        //this.handlePortal(); // Folia - region threading
         this.applyGravity();
         this.move(MoverType.SELF, this.getDeltaMovement());
         this.applyEffectsFromBlocks();
@@ -219,13 +219,21 @@ public class PrimedTnt extends Entity implements TraceableEntity {
     @Override
     public @Nullable Entity teleport(final TeleportTransition transition) {
         Entity newEntity = super.teleport(transition);
-        if (newEntity instanceof PrimedTnt tnt) {
-            tnt.setUsedPortal(true);
-        }
+        // Folia - region threading - move to post change
 
         return newEntity;
     }
 
+    // Folia start - region threading
+    @Override
+    public void postChangeDimension() {
+        super.postChangeDimension();
+        if (this instanceof PrimedTnt primedTnt) {
+            primedTnt.setUsedPortal(true);
+        }
+    }
+    // Folia end - region threading
+
     @Override
     public final boolean hurtServer(final ServerLevel level, final DamageSource source, final float damage) {
         return false;
