--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -111,10 +_,10 @@
     public static final Component DUPLICATE_LOGIN_DISCONNECT_MESSAGE = Component.translatable("multiplayer.disconnect.duplicate_login");
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final int SEND_PLAYER_INFO_INTERVAL = 600;
-    private static final SimpleDateFormat BAN_DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd 'at' HH:mm:ss z", Locale.ROOT);
+    private static final ThreadLocal<SimpleDateFormat> BAN_DATE_FORMAT = ThreadLocal.withInitial(() -> new SimpleDateFormat("yyyy-MM-dd 'at' HH:mm:ss z", Locale.ROOT)); // Folia - region threading - SDF is not thread-safe
     private final MinecraftServer server;
     public final List<ServerPlayer> players = new java.util.concurrent.CopyOnWriteArrayList(); // CraftBukkit - ArrayList -> CopyOnWriteArrayList: Iterator safety
-    private final Map<UUID, ServerPlayer> playersByUUID = Maps.newHashMap();
+    private final Map<UUID, ServerPlayer> playersByUUID = new java.util.concurrent.ConcurrentHashMap<>(); // Folia - region threading - change to CHM - Note: we do NOT expect concurrency PER KEY!
     private final UserBanList bans;
     private final IpBanList ipBans;
     private final ServerOpList ops;
@@ -135,6 +_,60 @@
     private final Map<String,ServerPlayer> playersByName = new java.util.HashMap<>();
     public @Nullable String collideRuleTeamName; // Paper - Configurable player collision
 
+    // Folia start - region threading
+    private final Object connectionsStateLock = new Object();
+    private final Map<String, Connection> connectionByName = new java.util.HashMap<>();
+    private final Map<UUID, Connection> connectionById = new java.util.HashMap<>();
+    private final it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<Connection> usersCountedAgainstLimit = new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<>();
+
+    public boolean pushPendingJoin(String userName, UUID byId, Connection conn) {
+        userName = userName.toLowerCase(java.util.Locale.ROOT);
+        Connection conflictingName, conflictingId;
+        synchronized (this.connectionsStateLock) {
+            conflictingName = this.connectionByName.get(userName);
+            conflictingId = this.connectionById.get(byId);
+
+            if (conflictingName == null && conflictingId == null) {
+                this.connectionByName.put(userName, conn);
+                this.connectionById.put(byId, conn);
+            }
+        }
+
+        Component message = Component.translatable("multiplayer.disconnect.duplicate_login", new Object[0]);
+
+        if (conflictingId != null || conflictingName != null) {
+            if (conflictingName != null && conflictingName.isPlayerConnected()) {
+                conflictingName.disconnectSafely(message, io.papermc.paper.connection.DisconnectionReason.DUPLICATE_LOGIN_MESSAGE);
+            }
+            if (conflictingName != conflictingId && conflictingId != null && conflictingId.isPlayerConnected()) {
+                conflictingId.disconnectSafely(message, io.papermc.paper.connection.DisconnectionReason.DUPLICATE_LOGIN_MESSAGE);
+            }
+        }
+
+        return conflictingName == null && conflictingId == null;
+    }
+
+    public void removeConnection(String userName, UUID byId, Connection conn) {
+        userName = userName.toLowerCase(java.util.Locale.ROOT);
+        synchronized (this.connectionsStateLock) {
+            this.connectionByName.remove(userName, conn);
+            this.connectionById.remove(byId, conn);
+            this.usersCountedAgainstLimit.remove(conn);
+        }
+    }
+
+    private boolean countConnection(Connection conn, int limit) {
+        synchronized (this.connectionsStateLock) {
+            int count = this.usersCountedAgainstLimit.size();
+            if (count >= limit) {
+                return false;
+            }
+            this.usersCountedAgainstLimit.add(conn);
+            return true;
+        }
+    }
+    // Folia end - region threading
+
     public PlayerList(
         final MinecraftServer server,
         final LayeredRegistryAccess<RegistryLayer> registries,
@@ -168,6 +_,10 @@
         String address = connection.getLoggableAddress(this.server.logIPs());
         LevelData levelData = level.getLevelData();
         ServerGamePacketListenerImpl playerConnection = new ServerGamePacketListenerImpl(this.server, connection, player, cookie);
+        // Folia start - rewrite login process
+        // only after setting the connection listener to game type, add the connection to this regions list
+        level.getCurrentWorldData().connections.add(connection);
+        // Folia end - rewrite login process
         connection.setupInboundProtocol(
             GameProtocols.SERVERBOUND_TEMPLATE.bind(RegistryFriendlyByteBuf.decorator(this.server.registryAccess()), playerConnection), playerConnection
         );
@@ -203,7 +_,7 @@
         this.sendPlayerPermissionLevel(player);
         player.getStats().markAllDirty();
         player.getRecipeBook().sendInitialRecipeBook(player);
-        this.updateEntireScoreboard(level.getScoreboard(), player);
+        //this.updateEntireScoreboard(level.getScoreboard(), player); // Folia - region threading
         this.server.invalidateStatus();
         MutableComponent component;
         if (player.getGameProfile().name().equalsIgnoreCase(oldName)) {
@@ -243,7 +_,7 @@
         this.cserver.getPluginManager().callEvent(playerJoinEvent);
 
         if (!player.connection.isAcceptingMessages()) {
-            return;
+            //return; // Folia - region threading - must still allow the player to connect, as we must add to chunk map before handling disconnect
         }
 
         final net.kyori.adventure.text.Component jm = playerJoinEvent.joinMessage();
@@ -258,8 +_,7 @@
         ClientboundPlayerInfoUpdatePacket packet = ClientboundPlayerInfoUpdatePacket.createPlayerInitializing(List.of(player)); // Paper - Add Listing API for Player
 
         final List<ServerPlayer> onlinePlayers = Lists.newArrayListWithExpectedSize(this.players.size() - 1); // Paper - Use single player info update packet on join
-        for (int i = 0; i < this.players.size(); ++i) {
-            ServerPlayer entityplayer1 = (ServerPlayer) this.players.get(i);
+        for (ServerPlayer entityplayer1 : this.players) { // Folia - region threading
 
             if (entityplayer1.getBukkitEntity().canSee(bukkitPlayer)) {
                 // Paper start - Add Listing API for Player
@@ -309,7 +_,7 @@
         // Paper start - Configurable player collision; Add to collideRule team if needed
         final net.minecraft.world.scores.Scoreboard scoreboard = this.getServer().getLevel(Level.OVERWORLD).getScoreboard();
         final PlayerTeam collideRuleTeam = scoreboard.getPlayerTeam(this.collideRuleTeamName);
-        if (this.collideRuleTeamName != null && collideRuleTeam != null && player.getTeam() == null) {
+        if (false && this.collideRuleTeamName != null && collideRuleTeam != null && player.getTeam() == null) { // Folia - region threading
             scoreboard.addPlayerToTeam(player.getScoreboardName(), collideRuleTeam);
         }
         // Paper end - Configurable player collision
@@ -414,7 +_,7 @@
 
     protected void save(final ServerPlayer player) {
         if (!player.getBukkitEntity().isPersistent()) return; // CraftBukkit
-        player.lastSave = MinecraftServer.currentTick; // Paper - Incremental chunk and player saving
+        player.lastSave = System.nanoTime(); // Folia - region threading - changed to nanoTime tracking
         this.playerIo.save(player);
         ServerStatsCounter stats = player.getStats(); // CraftBukkit
         if (stats != null) {
@@ -448,7 +_,7 @@
         // CraftBukkit end
 
         // Paper start - Configurable player collision; Remove from collideRule team if needed
-        if (this.collideRuleTeamName != null) {
+        if (false && this.collideRuleTeamName != null) { // Folia - region threading
             final net.minecraft.world.scores.Scoreboard scoreBoard = this.server.getLevel(Level.OVERWORLD).getScoreboard();
             final PlayerTeam team = scoreBoard.getPlayersTeam(this.collideRuleTeamName);
             if (player.getTeam() == team && team != null) {
@@ -496,6 +_,7 @@
 
         level.removePlayerImmediately(player, Entity.RemovalReason.UNLOADED_WITH_PLAYER);
         player.retireScheduler(); // Paper - Folia schedulers
+        player.getBukkitEntity().packetProcessor.close(); // Folia - region threading
         player.getAdvancements().clearTriggers();
         this.players.remove(player);
         this.playersByName.remove(player.getScoreboardName().toLowerCase(java.util.Locale.ROOT)); // Spigot
@@ -514,8 +_,7 @@
         // CraftBukkit start
         // this.broadcastAll(new ClientboundPlayerInfoRemovePacket(List.of(player.getUUID())));
         ClientboundPlayerInfoRemovePacket packet = new ClientboundPlayerInfoRemovePacket(List.of(player.getUUID()));
-        for (int i = 0; i < this.players.size(); i++) {
-            ServerPlayer otherPlayer = this.players.get(i);
+        for (ServerPlayer otherPlayer : this.players) { // Folia - region threading
 
             if (otherPlayer.getBukkitEntity().canSee(player.getBukkitEntity())) {
                 otherPlayer.connection.send(packet);
@@ -538,7 +_,7 @@
         }
     }
     // Paper end - PlayerLoginEvent
-    public LoginResult canPlayerLogin(final SocketAddress address, final NameAndId nameAndId) { // Paper - PlayerLoginEvent
+    public LoginResult canPlayerLogin(final SocketAddress address, final NameAndId nameAndId, final Connection connection) { // Paper - PlayerLoginEvent // Folia - region threading - add connection parameter
         LoginResult whitelistEventResult; // Paper
         // Paper start - Fix MC-158900
         UserBanListEntry ban1;
@@ -547,7 +_,7 @@
             // Paper end - Fix MC-158900
             MutableComponent reason = Component.translatable("multiplayer.disconnect.banned.reason", ban.getReasonMessage());
             if (ban.getExpires() != null) {
-                reason.append(Component.translatable("multiplayer.disconnect.banned.expiration", BAN_DATE_FORMAT.format(ban.getExpires())));
+                reason.append(Component.translatable("multiplayer.disconnect.banned.expiration", BAN_DATE_FORMAT.get().format(ban.getExpires()))); // Folia - region threading - SDF is not thread-safe
             }
 
             return new LoginResult(reason, org.bukkit.event.player.PlayerLoginEvent.Result.KICK_BANNED); // Paper - PlayerLoginEvent
@@ -560,12 +_,12 @@
                 IpBanListEntry ban = this.ipBans.get(address);
                 MutableComponent reason = Component.translatable("multiplayer.disconnect.banned_ip.reason", ban.getReasonMessage());
                 if (ban.getExpires() != null) {
-                    reason.append(Component.translatable("multiplayer.disconnect.banned_ip.expiration", BAN_DATE_FORMAT.format(ban.getExpires())));
+                    reason.append(Component.translatable("multiplayer.disconnect.banned_ip.expiration", BAN_DATE_FORMAT.get().format(ban.getExpires()))); // Folia - region threading - SDF is not thread-safe
                 }
 
                 return new LoginResult(reason, org.bukkit.event.player.PlayerLoginEvent.Result.KICK_BANNED); // Paper - PlayerLoginEvent
             } else {
-                return this.canBypassFullServerLogin(nameAndId, new LoginResult(Component.translatable("multiplayer.disconnect.server_full"), org.bukkit.event.player.PlayerLoginEvent.Result.KICK_FULL)); // Paper - PlayerServerFullCheckEvent
+                return this.canBypassFullServerLogin(nameAndId, new LoginResult(Component.translatable("multiplayer.disconnect.server_full"), org.bukkit.event.player.PlayerLoginEvent.Result.KICK_FULL), connection); // Paper - PlayerServerFullCheckEvent // Folia - region threading - add connection parameter
             }
         }
     }
@@ -585,15 +_,18 @@
             dupes.add(serverPlayer);
         }
 
-        for (ServerPlayer player : dupes) {
-            player.connection.disconnect(DUPLICATE_LOGIN_DISCONNECT_MESSAGE, io.papermc.paper.connection.DisconnectionReason.DUPLICATE_LOGIN_MESSAGE); // Paper - disconnect API
-        }
+        // Folia - region threading - rewrite login process - moved to pushPendingJoin
 
         return !dupes.isEmpty();
     }
 
     // Paper start - respawn event
     public ServerPlayer respawn(final ServerPlayer serverPlayer, final boolean keepAllPlayerData, final Entity.RemovalReason removalReason, final org.bukkit.event.player.PlayerRespawnEvent.RespawnReason respawnReason) {
+        // Folia start - region threading
+        if (true) {
+            throw new UnsupportedOperationException("Must use teleportAsync while in region threading");
+        }
+        // Folia end - region threading
         ServerPlayer.RespawnResult result = serverPlayer.findRespawnPositionAndUseSpawnBlock0(!keepAllPlayerData, TeleportTransition.DO_NOTHING, respawnReason);
         if (result == null) { // disconnected player during the respawn event
             return serverPlayer;
@@ -735,10 +_,10 @@
     public void tick() {
         if (++this.sendAllPlayerInfoIn > 600) {
             // CraftBukkit start
-            for (int i = 0; i < this.players.size(); ++i) {
-                final ServerPlayer target = this.players.get(i);
+            ServerPlayer[] players = this.players.toArray(new ServerPlayer[0]); // Folia - region threading
+            for (final ServerPlayer target : players) { // Folia - region threading
 
-                target.connection.send(new ClientboundPlayerInfoUpdatePacket(EnumSet.of(ClientboundPlayerInfoUpdatePacket.Action.UPDATE_LATENCY), com.google.common.collect.Collections2.filter(this.players, t -> target.getBukkitEntity().canSee(t.getBukkitEntity()))));
+                target.connection.send(new ClientboundPlayerInfoUpdatePacket(EnumSet.of(ClientboundPlayerInfoUpdatePacket.Action.UPDATE_LATENCY), com.google.common.collect.Collections2.filter(java.util.Arrays.asList(players),t -> target.getBukkitEntity().canSee(t.getBukkitEntity())))); // Folia - region threading
             }
             // CraftBukkit end
             this.sendAllPlayerInfoIn = 0;
@@ -756,8 +_,8 @@
     }
 
     public void broadcastAll(Packet packet, Level world) {
-        for (int i = 0; i < world.players().size(); ++i) {
-            ((ServerPlayer) world.players().get(i)).connection.send(packet);
+        for (net.minecraft.world.entity.player.Player player : world.players()) { // Folia - region threading
+            ((ServerPlayer) player).connection.send(packet); // Folia - region threading
         }
 
     }
@@ -794,8 +_,7 @@
         if (team == null) {
             this.broadcastSystemMessage(message, false);
         } else {
-            for (int i = 0; i < this.players.size(); i++) {
-                ServerPlayer targetPlayer = this.players.get(i);
+            for (ServerPlayer targetPlayer : this.players) { // Folia - region threading
                 if (targetPlayer.getTeam() != team) {
                     targetPlayer.sendSystemMessage(message);
                 }
@@ -804,10 +_,11 @@
     }
 
     public String[] getPlayerNamesArray() {
-        String[] names = new String[this.players.size()];
+        List<ServerPlayer> players = new java.util.ArrayList<>(this.players); // Folia - region threading
+        String[] names = new String[players.size()]; // Folia - region threading
 
-        for (int i = 0; i < this.players.size(); i++) {
-            names[i] = this.players.get(i).getGameProfile().name();
+        for (int i = 0; i < players.size(); i++) { // Folia - region threading
+            names[i] = players.get(i).getGameProfile().name(); // Folia - region threading
         }
 
         return names;
@@ -834,7 +_,9 @@
             );
         ServerPlayer player = this.getPlayer(nameAndId.id());
         if (player != null) {
-            this.sendPlayerPermissionLevel(player);
+            player.getBukkitEntity().taskScheduler.schedule((ServerPlayer serverPlayer) -> { // Folia - region threading
+            this.sendPlayerPermissionLevel(serverPlayer); // Folia - region threading
+            }, null, 1L); // Folia - region threading
         }
     }
 
@@ -842,7 +_,11 @@
         if (this.ops.remove(nameAndId)) {
             ServerPlayer player = this.getPlayer(nameAndId.id());
             if (player != null) {
-                this.sendPlayerPermissionLevel(player);
+                // Folia start - region threading
+                player.getBukkitEntity().taskScheduler.schedule((ServerPlayer serverPlayer) -> {
+                    this.sendPlayerPermissionLevel(serverPlayer);
+                }, null, 1L);
+                // Folia end - region threading
             }
         }
     }
@@ -872,8 +_,8 @@
     }
 
     // Paper start - whitelist verify event / login event
-    public LoginResult canBypassFullServerLogin(final NameAndId nameAndId, final LoginResult currentResult) {
-        final boolean shouldKick = this.players.size() >= this.getMaxPlayers() && !this.canBypassPlayerLimit(nameAndId);
+    public LoginResult canBypassFullServerLogin(final NameAndId nameAndId, final LoginResult currentResult, final Connection connection) { // Folia - region threading - add connection parameter
+        final boolean shouldKick = !this.countConnection(connection, this.getMaxPlayers()) && !this.canBypassPlayerLimit(nameAndId); // Folia - region threading - we control connection state here now async, not player list size
         final io.papermc.paper.event.player.PlayerServerFullCheckEvent fullCheckEvent = new io.papermc.paper.event.player.PlayerServerFullCheckEvent(
             new com.destroystokyo.paper.profile.CraftPlayerProfile(nameAndId),
             io.papermc.paper.adventure.PaperAdventure.asAdventure(currentResult.message),
@@ -932,8 +_,7 @@
         final ResourceKey<Level> dimension,
         final Packet<?> packet
     ) {
-        for (int i = 0; i < this.players.size(); i++) {
-            ServerPlayer player = this.players.get(i);
+        for (ServerPlayer player : this.players) { // Folia - region threading
             // CraftBukkit start - Test if player receiving packet can see the source of the packet
             if (except != null && !player.getBukkitEntity().canSee(except.getBukkitEntity())) {
                continue;
@@ -956,12 +_,18 @@
     }
     public void saveAll(final int interval) {
         // Paper end - Incremental chunk and player saving
-        io.papermc.paper.util.MCUtil.ensureMain("Save Players" , () -> { // Paper - Ensure main
+        //io.papermc.paper.util.MCUtil.ensureMain("Save Players" , () -> { // Paper - Ensure main // Folia - region threading
         // Paper start - Incremental chunk and player saving
             int numSaved = 0;
-            final long now = MinecraftServer.currentTick;
+            final long now = System.nanoTime(); // Folia - region threading
+            final long timeInterval = (long)interval * io.papermc.paper.threadedregions.TickRegionScheduler.TIME_BETWEEN_TICKS; // Folia - region threading
             for (final ServerPlayer player : this.players) {
-            if (interval == -1 || now - player.lastSave >= interval) {
+            // Folia start - region threading
+            if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(player)) {
+                continue;
+            }
+            // Folia end - region threading
+            if (interval == -1 || now - player.lastSave >= timeInterval) { // Folia - region threading
                 this.save(player);
                 if (interval != -1 && ++numSaved >= io.papermc.paper.configuration.GlobalConfiguration.get().playerAutoSave.maxPerTick()) {
                     break;
@@ -969,7 +_,7 @@
             }
             }
         // Paper end - Incremental chunk and player saving
-        return null; }); // Paper - ensure main
+        //return null; }); // Paper - ensure main // Folia - region threading
     }
 
     public UserWhiteList getWhiteList() {
@@ -1079,6 +_,20 @@
     }
 
     public void removeAll(boolean isRestarting) {
+        // Folia start - region threading
+        // just send disconnect packet, don't modify state
+        for (ServerPlayer player : this.players) {
+            final Component shutdownMessage = io.papermc.paper.adventure.PaperAdventure.asVanilla(this.server.server.shutdownMessage()); // Paper - Adventure
+            // CraftBukkit end
+
+            player.connection.send(new net.minecraft.network.protocol.common.ClientboundDisconnectPacket(shutdownMessage), net.minecraft.network.PacketSendListener.thenRun(() -> {
+                player.connection.connection.disconnect(shutdownMessage);
+            }));
+        }
+        if (true) {
+            return;
+        }
+        // Folia end - region threading
         // Paper end
         // CraftBukkit start - disconnect safely
         for (ServerPlayer player : this.players) {
@@ -1088,7 +_,7 @@
         // CraftBukkit end
 
         // Paper start - Configurable player collision; Remove collideRule team if it exists
-        if (this.collideRuleTeamName != null) {
+        if (false && this.collideRuleTeamName != null) { // Folia - region threading
             final net.minecraft.world.scores.Scoreboard scoreboard = this.getServer().getLevel(Level.OVERWORLD).getScoreboard();
             final PlayerTeam team = scoreboard.getPlayersTeam(this.collideRuleTeamName);
             if (team != null) scoreboard.removePlayerTeam(team);
