From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java b/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
index bebccbfcd65a8f5b9984684272368bbe11766c87..1cbfdaf981925c9862c8ead0e89535b915e1315c 100644
--- a/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
+++ b/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
@@ -54,7 +54,7 @@ public class FollowOwnerGoal extends Goal {
     public boolean canContinueToUse() {
         return !this.navigation.isDone()
             && !this.tamable.unableToMoveToOwner()
-            && !(this.tamable.distanceToSqr(this.owner) <= this.stopDistance * this.stopDistance);
+            && !(this.owner.level() == this.tamable.level() && this.tamable.distanceToSqr(this.owner) <= this.stopDistance * this.stopDistance); // Folia - region threading - check level
     }
 
     @Override
