--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -58,8 +_,19 @@
         this.setOwner(EntityReference.of(owner));
     }
 
+    // Folia start - region threading
+    // In general, this is an entire mess. At the time of writing, there are fifty usages of getOwner.
+    // Usage of this function is to avoid concurrency issues, even if it sacrifices behavior.
+    @Nullable
     @Override
-    public @Nullable Entity getOwner() {
+    public Entity getOwner() {
+        Entity ret = this.getOwnerRaw();
+        return ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(ret) && (ret == null || !ret.isRemoved()) ? ret : null;
+    }
+    // Folia end - region threading
+
+    public @Nullable Entity getOwnerRaw() { // Folia - region threading
+        if (!this.hasNullCallback()) ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this, "Cannot update owner state asynchronously"); // Folia - region threading
         return EntityReference.getEntity(this.owner, this.level());
     }
 
@@ -460,7 +_,7 @@
     @Override
     public boolean mayInteract(final ServerLevel level, final BlockPos pos) {
         Entity owner = this.getOwner();
-        return owner instanceof Player ? owner.mayInteract(level, pos) : owner == null || level.getGameRules().get(GameRules.MOB_GRIEFING);
+        return owner instanceof Player && ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(owner) ? owner.mayInteract(level, pos) : owner == null || level.getGameRules().get(GameRules.MOB_GRIEFING); // Folia - region threading
     }
 
     public boolean mayBreak(final ServerLevel level) {
