--- a/net/minecraft/world/level/block/EndGatewayBlock.java
+++ b/net/minecraft/world/level/block/EndGatewayBlock.java
@@ -115,6 +_,15 @@
             if (teleportPosition == null) {
                 return null;
             } else {
+                return getTeleportTransition(currentLevel, entity, teleportPosition); // Folia - region threading
+            }
+        } else {
+            return null;
+        }
+    }
+
+    // Folia start - region threading
+    public static TeleportTransition getTeleportTransition(final ServerLevel currentLevel, final Entity entity, final Vec3 teleportPosition) {
                 return entity instanceof ThrownEnderpearl
                     ? new TeleportTransition(currentLevel, teleportPosition, Vec3.ZERO, 0.0F, 0.0F, Set.of(), TeleportTransition.PLACE_PORTAL_TICKET, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.END_GATEWAY) // CraftBukkit
                     : new TeleportTransition(
@@ -127,11 +_,28 @@
                         TeleportTransition.PLACE_PORTAL_TICKET,
                         org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.END_GATEWAY // CraftBukkit
                     );
-            }
-        } else {
-            return null;
-        }
-    }
+    }
+
+    @Override
+    public boolean portalAsync(ServerLevel sourceWorld, Entity portalTarget, BlockPos portalPos) {
+        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(portalTarget)) {
+            return false;
+        }
+        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(sourceWorld, portalPos)) {
+            return false;
+        }
+
+        BlockEntity tile = sourceWorld.getBlockEntity(portalPos);
+
+        if (!(tile instanceof TheEndGatewayBlockEntity endGateway)) {
+            return false;
+        }
+
+        return TheEndGatewayBlockEntity.teleportRegionThreading(
+            sourceWorld, portalPos, portalTarget, endGateway, TeleportTransition.PLACE_PORTAL_TICKET
+        );
+    }
+    // Folia end - region threading
 
     @Override
     protected RenderShape getRenderShape(final BlockState state) {
