From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/entity/ai/village/poi/PoiManager.java b/net/minecraft/world/entity/ai/village/poi/PoiManager.java
index a67c3ccfd24172df1bd2e440a48d7e5f262af82f..1fa4c194c9055cbaac8a1806e896014588817d8b 100644
--- a/net/minecraft/world/entity/ai/village/poi/PoiManager.java
+++ b/net/minecraft/world/entity/ai/village/poi/PoiManager.java
@@ -60,11 +60,13 @@ public class PoiManager extends SectionStorage<PoiSection, PoiSection.Packed> im
     }
 
     private void updateDistanceTracking(long section) {
+        synchronized (this.villageDistanceTracker) { // Folia - region threading
         if (this.isVillageCenter(section)) {
             this.villageDistanceTracker.setSource(section, POI_DATA_SOURCE);
         } else {
             this.villageDistanceTracker.removeSource(section);
         }
+        } // Folia - region threading
     }
 
     @Override
@@ -420,10 +422,12 @@ public class PoiManager extends SectionStorage<PoiSection, PoiSection.Packed> im
     }
 
     public int sectionsToVillage(final SectionPos sectionPos) {
+        synchronized (this.villageDistanceTracker) { // Folia - region threading
         // Paper start - rewrite chunk system
         this.villageDistanceTracker.propagateUpdates();
         return convertBetweenLevels(this.villageDistanceTracker.getLevel(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkSectionKey(sectionPos)));
         // Paper end - rewrite chunk system
+        } // Folia - region threading
     }
 
     private boolean isVillageCenter(final long sectionPos) {
@@ -434,7 +438,9 @@ public class PoiManager extends SectionStorage<PoiSection, PoiSection.Packed> im
 
     @Override
     public void tick(final BooleanSupplier haveTime) {
+        synchronized (this.villageDistanceTracker) { // Folia - region threading
         this.villageDistanceTracker.propagateUpdates(); // Paper - rewrite chunk system
+        } // Folia - region threading
     }
 
     @Override
