From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/entity/ai/navigation/PathNavigation.java b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
index 1eb0afab932c3828a1f19861994b63b247df0fbb..bf39bc75b5b70b35d3064ebb5af37db7af07f453 100644
--- a/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -99,12 +99,12 @@ public abstract class PathNavigation {
     }
 
     public void recomputePath() {
-        if (this.level.getGameTime() - this.timeLastRecompute <= 20L || !this.canUpdatePath()) {
+        if (this.tick - this.timeLastRecompute <= 20L || !this.canUpdatePath()) { // Folia - region threading
             this.hasDelayedRecomputation = true;
         } else if (this.targetPos != null) {
             this.path = null;
             this.path = this.createPath(this.targetPos, this.reachRange);
-            this.timeLastRecompute = this.level.getGameTime();
+            this.timeLastRecompute = this.tick; // Folia - region threading
             this.hasDelayedRecomputation = false;
         }
     }
@@ -219,7 +219,7 @@ public abstract class PathNavigation {
 
     public boolean moveTo(final Entity target, final double speedModifier) {
         // Paper start - Perf: Optimise pathfinding
-        if (this.pathfindFailures > 10 && this.path == null && net.minecraft.server.MinecraftServer.currentTick < this.lastFailure + 40) {
+        if (this.pathfindFailures > 10 && this.path == null && this.tick < this.lastFailure + 40) { // Folia - region threading
             return false;
         }
         // Paper end - Perf: Optimise pathfinding
@@ -231,7 +231,7 @@ public abstract class PathNavigation {
             return true;
         } else {
             this.pathfindFailures++;
-            this.lastFailure = net.minecraft.server.MinecraftServer.currentTick;
+            this.lastFailure = this.tick; // Folia - region threading
             return false;
         }
         // Paper end - Perf: Optimise pathfinding
