From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/entity/projectile/FishingHook.java b/net/minecraft/world/entity/projectile/FishingHook.java
index 3e29bd82521b18a493f9b5b5b23cdb9c03cc94ba..fc91c859a5b37b030c7faab97e1e960dad6b5499 100644
--- a/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/net/minecraft/world/entity/projectile/FishingHook.java
@@ -270,6 +270,11 @@ public class FishingHook extends Projectile {
     }
 
     private boolean shouldStopFishing(final Player owner) {
+        // Folia start - region threading
+        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(owner)) {
+            return true;
+        }
+        // Folia end - region threading
         if (owner.canInteractWithLevel()) {
             ItemStack selectedItem = owner.getMainHandItem();
             ItemStack selectedItemOffHand = owner.getOffhandItem();
@@ -605,10 +610,18 @@ public class FishingHook extends Projectile {
 
     @Override
     public void remove(final Entity.RemovalReason reason, final org.bukkit.event.entity.EntityRemoveEvent.@Nullable Cause cause) { // CraftBukkit - add Bukkit remove cause
-        this.updateOwnerInfo(null);
+        //this.updateOwnerInfo(null); // Folia - region threading - move into preRemove
         super.remove(reason, cause); // CraftBukkit - add Bukkit remove cause
     }
 
+    // Folia start - region threading
+    @Override
+    protected void preRemove(RemovalReason reason) {
+        super.preRemove(reason);
+        this.updateOwnerInfo(null);
+    }
+    // Folia end - region threading
+
     @Override
     public void onClientRemoval() {
         this.updateOwnerInfo(null);
