From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/server/commands/AdvancementCommands.java b/net/minecraft/server/commands/AdvancementCommands.java
index 17b74b385c8e5204da8c692f8ebc4367d6aad5d9..5b735ea51fb3606555d3f9a5061af0a806f7e5f1 100644
--- a/net/minecraft/server/commands/AdvancementCommands.java
+++ b/net/minecraft/server/commands/AdvancementCommands.java
@@ -240,7 +240,12 @@ public class AdvancementCommands {
         int count = 0;
 
         for (ServerPlayer player : players) {
-            count += action.perform(player, advancements, showAdvancements);
+            // Folia start - region threading
+            count += 1;
+            player.getBukkitEntity().taskScheduler.scheduleOrExecute((ServerPlayer newPlayer) -> {
+                action.perform(newPlayer, advancements, showAdvancements);
+            });
+            // Folia end - region threading
         }
 
         if (count == 0) {
@@ -309,9 +314,12 @@ public class AdvancementCommands {
         }
 
         for (ServerPlayer player : players) {
-            if (action.performCriterion(player, holder, criterion)) {
-                count++;
-            }
+            // Folia start - region threading
+            ++count;
+            player.getBukkitEntity().taskScheduler.scheduleOrExecute((ServerPlayer newPlayer) -> {
+                action.performCriterion(newPlayer, holder, criterion);
+            });
+            // Folia end - region threading
         }
 
         if (count == 0) {
