From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/network/protocol/PacketUtils.java b/net/minecraft/network/protocol/PacketUtils.java
index 144c18c08bf8f03d907b97f968aa171a9fbf997f..be9fdca704391f8d77b87b4d75e67bb6597063c2 100644
--- a/net/minecraft/network/protocol/PacketUtils.java
+++ b/net/minecraft/network/protocol/PacketUtils.java
@@ -19,8 +19,16 @@ public class PacketUtils {
     }
 
     public static <T extends PacketListener> void ensureRunningOnSameThread(final Packet<T> packet, final T listener, final PacketProcessor packetProcessor) throws RunningOnDifferentThreadException {
-        if (!packetProcessor.isSameThread()) {
-            packetProcessor.scheduleIfPossible(listener, packet);
+        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThread()) { // Folia - region threading
+            // Folia start - region threading
+            if (listener instanceof net.minecraft.server.network.ServerGamePacketListenerImpl gamePacketListener) {
+                gamePacketListener.player.getBukkitEntity().schedulePacket(listener, packet);
+            } else if (listener instanceof net.minecraft.server.network.ServerConfigurationPacketListenerImpl || listener instanceof net.minecraft.server.network.ServerLoginPacketListenerImpl) {
+                io.papermc.paper.threadedregions.RegionizedServer.getInstance().schedulePacket(listener, packet);
+            } else {
+                throw new UnsupportedOperationException("Unknown listener: " + listener);
+            }
+            // Folia end - region threading
             throw RunningOnDifferentThreadException.RUNNING_ON_DIFFERENT_THREAD;
         }
     }
