From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/level/dimension/end/EnderDragonFight.java b/net/minecraft/world/level/dimension/end/EnderDragonFight.java
index 6cfaf75e92d194fe4ad1d404bc12ddec5ca18b8a..91447b6fcae1bb81ee5c86f76e00ad3d1e8ae446 100644
--- a/net/minecraft/world/level/dimension/end/EnderDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EnderDragonFight.java
@@ -80,7 +80,7 @@ public class EnderDragonFight extends SavedData {
     private Predicate<Entity> validPlayer;
     public ServerBossEvent dragonEvent;
     public ServerLevel level;
-    private BlockPos origin;
+    public BlockPos origin; // Folia - region threading - public
     public final List<Integer> gateways;
     private final BlockPattern exitPortalPattern = BlockPatternBuilder.start()
         .aisle("       ", "       ", "       ", "   #   ", "       ", "       ", "       ")
@@ -240,6 +240,12 @@ public class EnderDragonFight extends SavedData {
         }
 
         List<? extends EnderDragon> entities = this.level.getDragons();
+        // Folia start - region threading
+        // we do not want to deal with any dragons NOT nearby
+        entities.removeIf((EnderDragon dragon) -> {
+            return !ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(dragon);
+        });
+        // Folia end - region threading
         if (entities.isEmpty()) {
             this.dragonKilled = true;
         } else {
@@ -360,8 +366,14 @@ public class EnderDragonFight extends SavedData {
 
         ChunkPos chunkOrigin = ChunkPos.containing(this.origin);
 
-        for (int x = -8 + chunkOrigin.x(); x <= 8 + chunkOrigin.x(); x++) {
-            for (int z = 8 + chunkOrigin.z(); z <= 8 + chunkOrigin.z(); z++) {
+        // Folia start - region threading
+        final int regionRadiusChk = 8 + this.level.regioniser.regionSectionChunkSize;
+        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.level, chunkOrigin.x(), chunkOrigin.z(), regionRadiusChk)) {
+            return false;
+        }
+        // Folia end - region threading
+        for (int x = -8 + chunkOrigin.x(); x <= 8 + chunkOrigin.x(); x++) { // Folia - diff on change - use above for radius
+            for (int z = 8 + chunkOrigin.z(); z <= 8 + chunkOrigin.z(); z++) { // Folia - diff on change - use above for radius
                 ChunkAccess chunk = this.level.getChunk(x, z, ChunkStatus.FULL, false);
                 if (!(chunk instanceof LevelChunk)) {
                     return false;
@@ -530,6 +542,11 @@ public class EnderDragonFight extends SavedData {
     }
 
     public void onCrystalDestroyed(final EndCrystal crystal, final DamageSource source) {
+        // Folia start - region threading
+        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.level, this.origin)) {
+            return;
+        }
+        // Folia end - region threading
         if (this.respawnStage != null && this.respawnCrystals.contains(crystal)) {
             this.abortRespawnSequence();
         } else {
@@ -560,7 +577,7 @@ public class EnderDragonFight extends SavedData {
 
     public boolean tryRespawn(@Nullable final BlockPos placedEndCrystalPos) { // placedEndCrystalPos is null if the tryRespawn() call was not caused by a placed end crystal
         // Paper end - Perf: Do crystal-portal proximity check before entity lookup
-        if (this.dragonKilled && this.respawnStage == null) {
+        if (this.dragonKilled && this.respawnStage == null && ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.level, this.origin)) { // Folia - region threading
             BlockPos location = this.exitPortalLocation;
             if (location == null) {
                 LOGGER.debug("Tried to respawn, but need to find the portal first.");
