From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/entity/npc/wanderingtrader/WanderingTraderSpawner.java b/net/minecraft/world/entity/npc/wanderingtrader/WanderingTraderSpawner.java
index 4dea05ff63055803e7f41b729fdeca35879bc0d8..0b61a879c13474c730efeea059b1a6190067fd48 100644
--- a/net/minecraft/world/entity/npc/wanderingtrader/WanderingTraderSpawner.java
+++ b/net/minecraft/world/entity/npc/wanderingtrader/WanderingTraderSpawner.java
@@ -32,28 +32,27 @@ public class WanderingTraderSpawner implements CustomSpawner {
     private static final int NUMBER_OF_SPAWN_ATTEMPTS = 10;
     private final RandomSource random = RandomSource.create();
     private final SavedDataStorage savedDataStorage;
-    private int tickDelay;
-    private @Nullable WanderingTraderData traderData;
+    // Folia - moved to global data
 
     public WanderingTraderSpawner(final SavedDataStorage savedDataStorage) {
         this.savedDataStorage = savedDataStorage;
-        this.tickDelay = Integer.MIN_VALUE; // Paper - Add Wandering Trader spawn rate config options
-        this.traderData = null;
+        // Folia - moved to global data
     }
 
     @Override
     public void tick(final ServerLevel level, final boolean spawnEnemies) {
+        WanderingTraderData data = this.getTraderData(); // Folia - moved up
         // Paper start - Add Wandering Trader spawn rate config options
-        if (this.tickDelay == Integer.MIN_VALUE) {
-            this.tickDelay = level.paperConfig().entities.spawning.wanderingTrader.spawnMinuteLength;
-            WanderingTraderData data = this.getTraderData();
+        if (data.tickDelay == Integer.MIN_VALUE) { // Folia - moved to WanderingTraderData
+            data.tickDelay = level.paperConfig().entities.spawning.wanderingTrader.spawnMinuteLength; // Folia - moved to WanderingTraderData
+            // Folia - move up
             data.setSpawnDelay(level.paperConfig().entities.spawning.wanderingTrader.spawnDayLength);
             data.setSpawnChance(level.paperConfig().entities.spawning.wanderingTrader.spawnChanceMin);
         }
         if (level.getGameRules().get(GameRules.SPAWN_WANDERING_TRADERS)) {
-            if (this.tickDelay - 1 <= 0) { // Paper - Prevent tickDelay going below 0
-                this.tickDelay = level.paperConfig().entities.spawning.wanderingTrader.spawnMinuteLength;
-                WanderingTraderData data = this.getTraderData();
+            if (data.tickDelay - 1 <= 0) { // Paper - Prevent tickDelay going below 0 // Folia - moved to WanderingTraderData
+                data.tickDelay = level.paperConfig().entities.spawning.wanderingTrader.spawnMinuteLength; // Folia - moved to WanderingTraderData
+                // Folia - move up
                 int spawnDelay = data.spawnDelay() - level.paperConfig().entities.spawning.wanderingTrader.spawnMinuteLength;
                 data.setSpawnDelay(spawnDelay);
                 if (spawnDelay <= 0) {
@@ -68,20 +67,21 @@ public class WanderingTraderSpawner implements CustomSpawner {
                         }
                     }
                 }
-            } else { this.tickDelay--; } // Paper - Prevent tickDelay going below 0
+            } else { data.tickDelay--; } // Paper - Prevent tickDelay going below 0 // Folia - moved to WanderingTraderData
         }
     }
 
     private WanderingTraderData getTraderData() {
-        if (this.traderData == null) {
-            this.traderData = this.savedDataStorage.computeIfAbsent(WanderingTraderData.TYPE);
+        final io.papermc.paper.threadedregions.RegionizedWorldData worldData = io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegionizedWorldData(); // Folia - region threading
+        if (worldData.wanderingTraderData == null) { // Folia - region threading
+            worldData.wanderingTraderData = new net.minecraft.world.level.saveddata.WanderingTraderData(); // Folia - region threading
         }
 
-        return this.traderData;
+        return worldData.wanderingTraderData; // Folia - region threading
     }
 
     private boolean spawn(final ServerLevel level) {
-        Player player = level.getRandomPlayer();
+        Player player = level.getRandomLocalPlayer(); // Folia - region threading
         if (player == null) {
             return true;
         }
