From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/damagesource/CombatTracker.java b/net/minecraft/world/damagesource/CombatTracker.java
index 857d83574a1c3b87bc99e895c47c4c260833b150..a1957863d9c6674cee8ad8044f5b9fab7e1586ca 100644
--- a/net/minecraft/world/damagesource/CombatTracker.java
+++ b/net/minecraft/world/damagesource/CombatTracker.java
@@ -74,7 +74,7 @@ public class CombatTracker {
     private Component getMessageForAssistedFall(
         final Entity attackerEntity, final Component attackerName, final String messageWithItem, final String messageWithoutItem
     ) {
-        ItemStack attackerItem = attackerEntity instanceof LivingEntity livingEntity ? livingEntity.getMainHandItem() : ItemStack.EMPTY;
+        ItemStack attackerItem = attackerEntity instanceof LivingEntity livingEntity && ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(livingEntity) ? livingEntity.getMainHandItem() : ItemStack.EMPTY; // Folia - region threading
         return !attackerItem.isEmpty() && attackerItem.has(DataComponents.CUSTOM_NAME)
             ? Component.translatable(messageWithItem, this.mob.getDisplayName(), attackerName, attackerItem.getDisplayName())
             : Component.translatable(messageWithoutItem, this.mob.getDisplayName(), attackerName);
@@ -100,7 +100,7 @@ public class CombatTracker {
     }
 
     private static @Nullable Component getDisplayName(final @Nullable Entity entity) {
-        return entity == null ? null : entity.getDisplayName();
+        return entity == null || !ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity) ? null : entity.getDisplayName(); // Folia - region threading
     }
 
     public Component getDeathMessage() {
