--- a/net/minecraft/world/entity/animal/happyghast/HappyGhast.java
+++ b/net/minecraft/world/entity/animal/happyghast/HappyGhast.java
@@ -560,7 +_,7 @@
             happyGhastBb.maxZ + 1.0
         );
 
-        for (Player player : this.level().players()) {
+        for (Player player : this.level().getLocalPlayers()) { // Folia - region threading - Vanilla should do an AABB lookup but whatever
             if (!player.isSpectator()) {
                 Entity rootVehicle = player.getRootVehicle();
                 if (!(rootVehicle instanceof HappyGhast) && ghastDetectionBox.contains(rootVehicle.position())) {
