--- a/net/minecraft/world/level/dimension/end/EnderDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EnderDragonFight.java
@@ -80,7 +_,7 @@
     private Predicate<Entity> validPlayer;
     public ServerBossEvent dragonEvent;
     public ServerLevel level;
-    private BlockPos origin;
+    public BlockPos origin; // Folia - region threading - public
     public final List<Integer> gateways;
     private final BlockPattern exitPortalPattern = BlockPatternBuilder.start()
         .aisle("       ", "       ", "       ", "   #   ", "       ", "       ", "       ")
@@ -240,6 +_,12 @@
         }
 
         List<? extends EnderDragon> entities = this.level.getDragons();
+        // Folia start - region threading
+        // we do not want to deal with any dragons NOT nearby
+        entities.removeIf((EnderDragon dragon) -> {
+            return !ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(dragon);
+        });
+        // Folia end - region threading
         if (entities.isEmpty()) {
             this.dragonKilled = true;
         } else {
@@ -360,8 +_,14 @@
 
         ChunkPos chunkOrigin = ChunkPos.containing(this.origin);
 
-        for (int x = -8 + chunkOrigin.x(); x <= 8 + chunkOrigin.x(); x++) {
-            for (int z = 8 + chunkOrigin.z(); z <= 8 + chunkOrigin.z(); z++) {
+        // Folia start - region threading
+        final int regionRadiusChk = 8 + this.level.regioniser.regionSectionChunkSize;
+        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.level, chunkOrigin.x(), chunkOrigin.z(), regionRadiusChk)) {
+            return false;
+        }
+        // Folia end - region threading
+        for (int x = -8 + chunkOrigin.x(); x <= 8 + chunkOrigin.x(); x++) { // Folia - diff on change - use above for radius
+            for (int z = 8 + chunkOrigin.z(); z <= 8 + chunkOrigin.z(); z++) { // Folia - diff on change - use above for radius
                 if (!(this.level.getChunk(x, z, ChunkStatus.FULL, false) instanceof LevelChunk levelChunk)) {
                     return false;
                 }
@@ -529,6 +_,11 @@
     }
 
     public void onCrystalDestroyed(final EndCrystal crystal, final DamageSource source) {
+        // Folia start - region threading
+        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.level, this.origin)) {
+            return;
+        }
+        // Folia end - region threading
         if (this.respawnStage != null && this.respawnCrystals.stream().anyMatch(ref -> ref.matches(crystal))) {
             this.abortRespawnSequence();
         } else {
@@ -559,7 +_,7 @@
 
     public boolean tryRespawn(@Nullable final BlockPos placedEndCrystalPos) { // placedEndCrystalPos is null if the tryRespawn() call was not caused by a placed end crystal
         // Paper end - Perf: Do crystal-portal proximity check before entity lookup
-        if (this.dragonKilled && this.respawnStage == null) {
+        if (this.dragonKilled && this.respawnStage == null && ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.level, this.origin)) { // Folia - region threading
             BlockPos location = this.exitPortalLocation;
             if (location == null) {
                 LOGGER.debug("Tried to respawn, but need to find the portal first.");
