From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/entity/monster/Vex.java b/net/minecraft/world/entity/monster/Vex.java
index e23aa0c86cd4070628f29b51eb4f804bb63a4f27..d9d2ac4e80a2257530daf7304750e152ae3c4938 100644
--- a/net/minecraft/world/entity/monster/Vex.java
+++ b/net/minecraft/world/entity/monster/Vex.java
@@ -341,7 +341,7 @@ public class Vex extends Monster implements TraceableEntity {
         @Override
         public void tick() {
             BlockPos boundOrigin = Vex.this.getBoundOrigin();
-            if (boundOrigin == null) {
+            if (boundOrigin == null || !ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor((net.minecraft.server.level.ServerLevel)Vex.this.level(), boundOrigin)) { // Folia - region threading
                 boundOrigin = Vex.this.blockPosition();
             }
 
