From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index 8b194a59f79eea32c9e7a46227034942f6a6e816..bb27a86b16daf6f026ec6acf7a6419ecd46741ef 100644
--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -344,24 +344,19 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
         if (!this.connection.isConnected()) {
             return;
         }
-        if (!this.cserver.isPrimaryThread()) {
-            org.bukkit.craftbukkit.util.Waitable waitable = new org.bukkit.craftbukkit.util.Waitable() {
-                @Override
-                protected @Nullable Object evaluate() {
-                    ServerCommonPacketListenerImpl.this.disconnect(details);
-                    return null;
-                }
-            };
-
-            this.server.processQueue.add(waitable);
-
-            try {
-                waitable.get();
-            } catch (InterruptedException e) {
-                Thread.currentThread().interrupt();
-            } catch (java.util.concurrent.ExecutionException e) {
-                throw new RuntimeException(e);
-            }
+        // Folia start - region threading
+        boolean onMain;
+        if (this instanceof ServerGamePacketListenerImpl serverGamePacketListener) {
+            onMain = ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(serverGamePacketListener.player);
+        } else if (this instanceof ServerConfigurationPacketListenerImpl configurationPacketListener) {
+            net.minecraft.server.level.ServerPlayer player = configurationPacketListener.switchToMain;
+            onMain = player == null ? io.papermc.paper.threadedregions.RegionizedServer.isGlobalTickThread() : ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(player);
+        } else {
+            onMain = io.papermc.paper.threadedregions.RegionizedServer.isGlobalTickThread();
+        }
+        if (!onMain) {
+            this.connection.disconnectSafely(details); // it HAS to be delayed/async to avoid deadlock if we try to wait for another region
+            // Folia end - region threading
             return;
         }
 
