From 8a1f7b93bacac28f423d2cf4b4c296608196ccc1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] folia File Patches


diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index ea144213f5ac2f813c9007113e4066a3295a7f76..97b35ede0023000f293595a2b349ec9d1c34ec5a 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -247,8 +247,20 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
         return this.sensing;
     }
 
+    // Folia start - region threading
+    public LivingEntity getTargetRaw() {
+        return this.target;
+    }
+    // Folia end - region threading
+
     @Override
     public @Nullable LivingEntity getTarget() {
+        // Folia start - region threading
+        if (this.target != null && (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.target) || this.target.isRemoved())) {
+            this.target = null;
+            return null;
+        }
+        // Folia end - region threading
         return this.asValidTarget(this.target);
     }
 
@@ -257,6 +269,11 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     }
 
     protected @Nullable LivingEntity asValidTarget(final @Nullable LivingEntity target) {
+        // Folia start - region threading
+        if (target != null && (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(target) || target.isRemoved())) {
+            return null;
+        }
+        // Folia end - region threading
         if (target instanceof Player player && (player.isCreative() || player.isSpectator())) {
             return null;
         } else {
@@ -1626,6 +1643,20 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
         }
     }
 
+    // Folia start - moved from removeAfterChangingDimensions()
+    @Override
+    protected void postRemoveAfterChangingDimensions() {
+        super.postRemoveAfterChangingDimensions();
+
+        for (EquipmentSlot slot : EquipmentSlot.VALUES) {
+            ItemStack itemStack = this.getItemBySlot(slot);
+            if (!itemStack.isEmpty()) {
+                itemStack.setCount(0);
+            }
+        }
+    }
+    // Folia end - moved from removeAfterChangingDimensions()
+
     @VisibleForTesting
     public void removeFreeWill() {
         this.removeAllGoals(goal -> true);
@@ -1640,12 +1671,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     protected void removeAfterChangingDimensions() {
         super.removeAfterChangingDimensions();
 
-        for (EquipmentSlot slot : EquipmentSlot.VALUES) {
-            ItemStack itemStack = this.getItemBySlot(slot);
-            if (!itemStack.isEmpty()) {
-                itemStack.setCount(0);
-            }
-        }
+        // Folia - move to postRemoveAfterChangingDimensions()
     }
 
     @Override
